File: samples/vdsp_op/custom_op/brightness/brightness.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "brightness_op.hpp"
#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input file", false,
                        "../data/images/dog.jpg");
  args.add<std::string>("output_file", '\0', "output file", false,
                        "brightness_op_result.jpg");
  args.add<std::string>("elf_file", '\0', "elf_file path", false,
                        "/opt/vastai/vaststreamx/data/elf/brightness");
  args.add<float>("scale", '\0', "brightness scale coefficient", false, 2.2);
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  uint32_t device_id = args.get<uint32_t>("device_id");
  float scale = args.get<float>("scale");

  vsx::SetDevice(device_id);
  auto brighness_op =
      vsx::BrightnessOp("img_brightness_adjust",
                        args.get<std::string>("elf_file"), device_id, scale);
  vsx::Image input_image;
  vsx::MakeVsxImage(args.get<std::string>("input_file"), input_image,
                    vsx::YUV_NV12);
  auto output_vacc = brighness_op.Process(input_image);
  // copy result to cpu memory
  auto output_cpu = vsx::Image(output_vacc.Format(), output_vacc.Width(),
                               output_vacc.Height());
  output_cpu.CopyFrom(output_vacc);

  cv::Mat out_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output_cpu, out_mat);
  cv::imwrite(args.get<std::string>("output_file"), out_mat);
  std::cout << "Write result to " << args.get<std::string>("output_file")
            << std::endl;
  return 0;
}

File: samples/vdsp_op/custom_op/brightness/brightness.py
import vaststreamx as vsx
import os
import sys
import cv2
import argparse
import numpy as np
from brightness_op import BrightnessOp

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../../../")
sys.path.append(common_path)

import common.utils as utils


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/dog.jpg",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="./custom_op_result.jpg",
        help="output file",
    )
    parser.add_argument(
        "--elf_file",
        default="/opt/vastai/vaststreamx/data/elf/brightness",
        help="elf file",
    )
    parser.add_argument(
        "--scale",
        default=2.2,
        type=float,
        help="brightness scale coefficient",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()

    brightness_op = BrightnessOp(
        "img_brightness_adjust", args.elf_file, args.device_id, args.scale
    )

    bgr888 = cv2.imread(args.input_file)
    yuv_nv12 = utils.cv_bgr888_to_nv12(bgr888)

    out_nv12 = brightness_op.process(yuv_nv12)
    cv_nv12 = vsx.as_numpy(out_nv12).squeeze()
    out_bgr888 = utils.cv_nv12_to_bgr888(cv_nv12)

    cv2.imwrite(args.output_file, out_bgr888)
    print("Write result to ", args.output_file)
